include/envoy/registry/registry.h

```diff
@@ -73,10 +83,17 @@ template <class T, class Base> class RegisterFactory {
   /**
    * Contructor that registers an instance of the factory with the FactoryRegistry.
    */
-  RegisterFactory() {
-    static T* instance = new T;
-    FactoryRegistry<Base>::registerFactory(*instance);
-  }
+  RegisterFactory() { FactoryRegistry<Base>::registerFactory(instance_); }
+
+  /**
+   * Destructor that removes an instance of the factory from the FactoryRegistry.
+   */
+  ~RegisterFactory() { FactoryRegistry<Base>::unregisterFactory(instance_); }
```

Comment:
This new setup creates a slightly bigger violation of the static ordering fiasco - we now have destructors being called in some unknown ordering (in addition to the constructors).  @akonradi mentioned this was done so that tests don't leak state.  Maybe it would be preferable to create an `unregister` method on the `FactoryRegistry` that manually removes the pointer (and allows the user to delete, maybe by returning as a unique_ptr) rather than doing it by default on destruction since we'd only need this in tests? @htuch thoughts?

Reply:
@akonradi can you point at the specific test where the leak was an issue?

Reply:
https://github.com/akonradi/envoy/blob/915c7d3b432ae9a4abbc4f69da5bf19694feb59a/test/common/network/resolver_impl_test.cc#L145 is where there could have been one if the resolver from https://github.com/akonradi/envoy/blob/915c7d3b432ae9a4abbc4f69da5bf19694feb59a/test/common/network/resolver_impl_test.cc#L106 was allowed to leak. FWIW it seems like we've avoided this in the rest of the tests because we don't register test factories and stick to testing the included ones.

Reply:
This is where a singleton override for test would be useful. @alyssawilk any thoughts on how far along the work is to provide this?

Reply:
Hasn't started (not assigned) but it looked like a pretty simple implementation.  If it's as few lines of code as I think it could be folded in here

Reply:
You're talking about #1808? I'd rather make that a separate pull request and either wait to merge it once it lands in master or add a TODO here.

Reply:
OK, TODO sounds good to me, and update the issue to reference, thanks.